### PR TITLE
Refactor RegexProperty to use a property type.

### PR DESCRIPTION
**Pull request by sharief007**
### Change Log
- Created class `RegexType` extending from `PropertyAndLeafType`.
- Conversion logic moved from the `RegexProperty` to the new type.
- Changed `RegexProperty` to extend `TypeBasedProperty` and to use the `RegexType`.
- Added `RegexType` to the list of default types in the leaf value handler.
- Fixed broken unit test cases because of the above changes.
- Added unit test cases for `RegexType`.

### CODE REVIEW

File: src/main/java/ch/jalu/configme/properties/types/RegexType.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * Constructor. Use {@link RegexType#REGEX} for the standard behaviour.
+     */
+    protected RegexType() {
+        super(Pattern.class);
+    }
+
+    @Override
+    public @Nullable Pattern convert(@Nullable Object object, @NotNull ConvertErrorRecorder errorRecorder) {
+        if (object != null && object instanceof String) {
```
Comment (ljacqu) [inline src/main/java/ch/jalu/configme/properties/types/RegexType.java:30]:
```suggestion
        if (object instanceof String) {
```

File: src/main/java/ch/jalu/configme/properties/types/RegexType.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        return null;
+    }
+
+    @Override
+    public @Nullable Object toExportValue(@NotNull Pattern value) {
+        return value.pattern();
+    }
+
+    @Override
+    public boolean canConvertToType(@NotNull TypeInfo typeInformation) {
```
Comment (ljacqu) [inline src/main/java/ch/jalu/configme/properties/types/RegexType.java:46]: No need to override this method—the implementation from `PropertyAndLeafType` does exactly what it needs :smile:

File: src/main/java/ch/jalu/configme/properties/types/RegexType.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Property type and mapper leaf type for regex.
+ */
+public class RegexType extends PropertyAndLeafType<Pattern> {
+    
+    /** Default regex type. */
+    public static final RegexType REGEX = new RegexType();
+
+    /**
+     * Constructor. Use {@link RegexType#REGEX} for the standard behaviour.
```
Comment (ljacqu) [inline src/main/java/ch/jalu/configme/properties/types/RegexType.java:22]:
```suggestion
     * Constructor. Use {@link RegexType#REGEX} for the standard behavior.
```

**Review by ljacqu — COMMENTED** [review]
Wonderful! Thank you for this PR. :tada:

I've left two small comments. If you don't have time to do them, I'm also happy to merge & apply changes myself.

**Comment by sharief007** [comment]
Applied changes as per the review comments.

**Comment by ljacqu** [comment]
Great work, thanks again!

**Pull request merged**